packages/flutter/lib/src/painting/image_stream.dart

```diff
@@ -2,35 +2,139 @@
+    bool returnValue = false;
+    assert(() {
+      returnValue = _debugDisposed;
+      return true;
+    }());
+    return returnValue;
+  }
+
+  /// Called by [ImageInfo.obtainImageHandle].
+  void _ref() {
+    assert(!_debugDisposed, 'Object has been disposed.');
+    _refCount += 1;
+  }
+
+  /// Release this image handle.
+  ///
+  /// Once all outstanding image handles have been disposed, the underlying
+  /// [ui.Image] will be disposed as well. Further attempts to use this handle
+  /// will throw.
+  ///
+  /// Clients must call this method once and only once for each call to
+  /// [ImageInfo.obtainImageHandle].
+  void dispose() {
+    assert(!_debugDisposed, 'Object has been disposed');
+    _refCount -= 1;
+    if (_refCount == 0) {
+      assert(() {
+        _debugDisposed = true;
+        return true;
+      }());
+      _image.dispose();
+    }
+  }
+}
+
 /// A [dart:ui.Image] object with its corresponding scale.
 ///
 /// ImageInfo objects are used by [ImageStream] objects to represent the
 /// actual data of the image once it has been obtained.
+///
+/// Before accessing the [image], callers must use [obtainImageHandle] to get
+/// an [ImageHandle]. When the caller no longer needs the image, they should
+/// call [ImageHandle.dispose].
 @immutable
 class ImageInfo {
   /// Creates an [ImageInfo] object for the given [image] and [scale].
   ///
-  /// Both the image and the scale must not be null.
+  /// The [image] and [scale] parameters must not be null.
   ///
-  /// The tag may be used to identify the source of this image.
-  const ImageInfo({ required this.image, this.scale = 1.0, this.debugLabel })
-    : assert(image != null),
-      assert(scale != null);
+  /// A client of this class must call [obtainImageHandle] before using the
+  /// [image], and then call [ImageHandle.dispose] when they will no longer need
+  /// the [image]. Once all clients have disposed of their handles, the [image]
+  /// will be in a disposed state as well and will no longer be usable.
+  ///
+  /// The [debugLabel] may be used to identify the source of this image.
+  ImageInfo({
+    required ui.Image image,
+    this.scale = 1.0,
+    this.debugLabel,
+  }) : assert(image != null),
+       assert(scale != null),
+       _imageHandle = ImageHandle._create(image);
 
   /// The raw image pixels.
   ///
   /// This is the object to pass to the [Canvas.drawImage],
   /// [Canvas.drawImageRect], or [Canvas.drawImageNine] methods when painting
   /// the image.
-  final ui.Image image;
+  ///
+  /// Do not call dispose on this object. Other clients of an
+  /// [ImageStream] may be using it. Instead, call [obtainImageHandle] and call
+  /// [ImageHandle.dispose] when the client no longer will use the image.
+  ui.Image get image {
+    assert(!_imageHandle.debugDisposed, 'Object has been disposed.');
+    assert(_imageHandle._refCount > 0, 'Call `obtainImageHandle` before accessing the image in an ImageInfo object.');
```

Comment:
`_refCount > 0` could mean that somebody else (not the current caller of `image`) has a valid handle, no? I guess, we can't really assert that the current caller has a valid handle...

(see my other comment above): That's why I wish we could have an API where you access the image from a valid handler...

Reply:
Yes - it could be someone else keeping the refcount up, not the caller.